docs/specifications/4-standard-compliance-suite.md

```diff
@@ -0,0 +1,500 @@
+the *jsii specification* was initially written. They are going to be gradually
+replaced by more focused tests with better descriptions.
+
+#### Type Unions are correctly disambiguated by the Kernel
+In certain cases, two or more types in a *Type Union* can overlap in such a way
+that they are all valid structural types for the value. Statically typed
+languages however will not be satisfied with structural typing only, and require
+the correct declared type to be preserved.
+
+<details><summary>Show test</summary>
+
+##### Reference Implementation
+```ts
+// GIVEN
+export interface BluePill {
+  readonly offeredTo: string;
+  readonly makesYouForgetTheMatrix?: boolean;
+}
+export interface RedPill {
+  readonly offeredTo: string;
+  readonly makesYouExitTheMatrix?: boolean;
+}
+export class Morpheus {
+  public static isBlue(pill: BluePill | RedPill): pill is BluePill {
+    const keys = new Set(Object.keys(pill));
+    switch (keys.size) {
+      case 1:
+        return keys.has('offeredTo');
+      case 2:
+      return keys.has('offeredTo') && keys.has('makesYouForgetTheMatrix');
+      default:
+        return false;
+    }
+  }
+  public static isRed(pill: BluePill | RedPill): pill is RedPill {
+    const keys = new Set(Object.keys(pill));
+    switch (keys.size) {
+      case 1:
+        return keys.has('offeredTo');
+      case 2:
+        return keys.has('offeredTo') && keys.has('makesYouExitTheMatrix');
+      default:
+        return false;
+    }
+  }
+  private constructor(){}
+}
+export class Neo {
+  public readonly tookBlue: boolean;
+  public readonly tookRed: boolean;
+
+  public constructor(public readonly pill: BluePill | RedPill) {
+    this.tookBlue = pill.offeredTo == 'Neo' && Morpheus.isBlue(pill);
+    this.tookRed = pill.offeredTo == 'Neo' && Morpheus.isRed(pill);
+  }
+}
+
+// WHEN
+const bluePillA = new Neo({ offeredTo: 'not Neo' });
+const bluePillB = new Neo({ offeredTo: 'Neo', makesYouForgetTheMatrix: true });
+const redPillA = new Neo({ offeredTo: 'not Neo' });
+const redPillB = new Neo({ offeredTo: 'Neo', makesYouExitTheMatrix: true });
+
+// THEN
+expect(bluePillA.pill instanceof BluePill).toBeTruthy();
+expect(bluePillA.tookBlue).toBeFalsy();
+expect(bluePillA.tookRed).toBeFalsy();
+
+expect(bluePillB.pill instanceof BluePill).toBeTruthy();
+expect(bluePillA.tookBlue).toBeTruthy();
+expect(bluePillA.tookRed).toBeFalsy();
+
+expect(redPillA.pill instanceof RedPill).toBeTruthy();
+expect(bluePillA.tookBlue).toBeFalsy();
+expect(bluePillA.tookRed).toBeFalsy();
+
+expect(redPillB.pill instanceof RedPill).toBeTruthy();
+expect(bluePillA.tookBlue).toBeFalsy();
+expect(bluePillA.tookRed).toBeTruthy();
+```
+
+##### Kernel Trace
+```
+```
+</details>
+
+#### Partially initialized object consumption
+When a constructor passes `this` out from **JavaScript** to the *host* app, the
+reference must be correctly identified and passed across.
+
+> :construction: The .NET Runtime does not currently honor object identity,
```

Comment:
That is interesting. Means that jsii needs a notion of what it means to be "the same object". I believe because of anonymous objects being returned under 2 different interfaces, we must also emit 2 proxies in general (Java etc) and so it just might be that comparing objects by identity using `==` is explicitly disallowed by the JSII spec. 

It might be worthwhile adding a `JSII.isSameObject(x, y)` static to the runtimes then to fill that gap, for example.

Reply:
That's a great idea and I'm stealing it.